modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +184,99 @@ private void ensureOpen() {
         }
-        final Supplier<String> messageSupplier = () -> format(
-            """
-                failed reading [%s/%s] at offset [%s]; this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
-                [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
-                retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
-            blobStore.bucket(),
-            blobKey,
-            start + currentOffset,
-            attempt,
-            currentStreamProgress,
-            failuresAfterMeaningfulProgress,
-            maxAttempts
+        logger.log(
+            level,
+            () -> format(
+                """
+                    failed %s [%s/%s] at offset [%s] with purpose [%s]; \
+                    this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
+                    [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
+                    retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
+                action,
+                blobStore.bucket(),
+                blobKey,
+                start + currentOffset,
+                purpose.getKey(),
+                attempt - 1,
+                currentStreamProgress,
+                failuresAfterMeaningfulProgress,
+                attempt - 1 - failuresAfterMeaningfulProgress
+            ),
+            e
         );
-        if (attempt >= maxAttempts + failuresAfterMeaningfulProgress) {
-            final var finalException = addSuppressedExceptions(e);
-            logger.warn(messageSupplier, finalException);
-            throw finalException;
+    }
+
+    private boolean shouldRetry() {
+        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
+            return false;
         }
-        logger.debug(messageSupplier, e);
-        attempt += 1;
-        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
-            failures.add(e);
+        if (purpose == OperationPurpose.INDICES) {
+            return true;
+        }
+        return attempt < maxAttempts + failuresAfterMeaningfulProgress;
+    }
+
+    private void maybeDelaySafely() {
+        final long delayInMillis = getRetryDelayInMillis();
+        if (delayInMillis > 0) {
+            try {
+                Thread.sleep(delayInMillis);
+            } catch (InterruptedException e) {
+                logger.info("delay interrupted", e);
+            }
+        }
+    }
+
+    // for testing
+    protected long getRetryDelayInMillis() {
+        final long initialDelayInMillis = 10L;
+        if (attempt > maxAttempts) {
+            // Cap max delay at 10 * 1024 millis, i.e. it retries every ~10 seconds at a minimum
+            return initialDelayInMillis << (Math.min(attempt - maxAttempts - 1, 10));
```

Comment:
The choice of initial and max delay is modelled after the ones in [IndicesService](https://github.com/elastic/elasticsearch/blob/d47fee26644ad93a75c0a83a62c3eee14d349d0f/server/src/main/java/org/elasticsearch/indices/IndicesService.java#L1375-L1376)